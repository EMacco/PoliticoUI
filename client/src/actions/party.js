import axios from 'axios';
import { GET_ERRORS, SET_POLITICAL_PARTIES } from '@actions/types';
import { setCurrentUser } from '@actions/authActions';
import setAuthToken from '../utils/setAuthToken';

const fetchPartyByID = (id, completionHandler) => {
  axios
    .get(`/parties/${id}`)
    .then(res => {
      completionHandler(res);
    })
    .catch(err => {
      dispatch({ type: GET_ERRORS, payload: { global: err.response.data.error } });
    });
};

const fetchAllParties = completionHandler => {
  axios
    .get(`/parties`)
    .then(res => {
      completionHandler(res);
    })
    .catch(err => {
      dispatch({ type: GET_ERRORS, payload: { global: err.response.data.error } });
    });
};

const changeUserParty = partyId => dispatch => {
  axios
    .post(`/parties/join`, { partyId })
    .then(res => {
      const { token, user } = res.data.data[0];
      localStorage.setItem('jwtToken', token);
      setAuthToken(token);
      dispatch(setCurrentUser(user));
    })
    .catch(err => {
      console.log(err);
      dispatch({ type: GET_ERRORS, payload: { global: err.response.data.error } });
    });
};

const fetchPoliticalParties = partyid => dispatch => {
  let parties = [];
  fetchAllParties(partiesRes => {
    const { data } = partiesRes.data;

    for (let ind = 0; ind < data.length; ind += 1) {
      const partyObject = {
        partyId: data[ind].id,
        partyLogo: data[ind].logourl,
        partyName: data[ind].name,
        hqAddress: data[ind].hqaddress
      };

      parties.push(partyObject);
      if (ind === data.length - 1) {
        dispatch({ type: SET_POLITICAL_PARTIES, payload: parties });
      }
    }
  });
};

export { fetchPartyByID, fetchAllParties, fetchPoliticalParties, changeUserParty };
