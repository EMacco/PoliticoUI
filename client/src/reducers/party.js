import { SET_POLITICAL_PARTIES } from '../actions/types';

const initialState = {
  parties: []
};

const partyReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_POLITICAL_PARTIES:
      return { ...state, parties: action.payload };
    default:
      return state;
  }
};

export default partyReducer;
