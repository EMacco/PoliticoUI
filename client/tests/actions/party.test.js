import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { fetchPoliticalParties, changeUserParty } from '@actions/party';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const store = mockStore({
  auth: {},
  dashboard: {},
  parties: {},
  offices: {}
});

describe('Offices actions', () => {
  it('should dispatch SET_POLITICAL_PARTIES on successful fetch', async () => {
    await store.dispatch(fetchPoliticalParties(1));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should dispatch SET_CURRENT_USER on successful change of party', async () => {
    await store.dispatch(changeUserParty(1));
    expect(store.getActions()).toMatchSnapshot();
  });
});
