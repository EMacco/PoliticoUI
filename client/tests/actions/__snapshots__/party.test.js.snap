// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offices actions should dispatch SET_CURRENT_USER on successful change of party 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "hqAddress": "festac lagos",
        "partyId": 1,
        "partyLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551452104/dggfgdaqi5demtki1dnt.png",
        "partyName": "peoples democratic party",
      },
      Object {
        "hqAddress": "festac lagos",
        "partyId": 2,
        "partyLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551452220/ukhkvobjnabok4ymt7g3.jpg",
        "partyName": "all progressives congress",
      },
      Object {
        "hqAddress": "lekki, lagos",
        "partyId": 3,
        "partyLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551466591/aou6cohppt8yckombbsm.png",
        "partyName": "all progressives grand alliance",
      },
      Object {
        "hqAddress": "google, universal",
        "partyId": 4,
        "partyLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1557735076/ogpffpwqi1s8hsuewlah.jpg",
        "partyName": "node package manager",
      },
    ],
    "type": "SET_POLITICAL_PARTIES",
  },
  Object {
    "payload": Object {
      "email": "testing@testing.com",
      "firstname": "John",
      "id": 1,
      "lastname": "Doe",
      "partyid": 1,
      "password": "password",
      "phone": "08088293829",
    },
    "type": "SET_CURRENT_USER",
  },
]
`;

exports[`Offices actions should dispatch SET_POLITICAL_PARTIES on successful fetch 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "hqAddress": "festac lagos",
        "partyId": 1,
        "partyLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551452104/dggfgdaqi5demtki1dnt.png",
        "partyName": "peoples democratic party",
      },
      Object {
        "hqAddress": "festac lagos",
        "partyId": 2,
        "partyLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551452220/ukhkvobjnabok4ymt7g3.jpg",
        "partyName": "all progressives congress",
      },
      Object {
        "hqAddress": "lekki, lagos",
        "partyId": 3,
        "partyLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1551466591/aou6cohppt8yckombbsm.png",
        "partyName": "all progressives grand alliance",
      },
      Object {
        "hqAddress": "google, universal",
        "partyId": 4,
        "partyLogo": "https://res.cloudinary.com/dn4pokov0/image/upload/v1557735076/ogpffpwqi1s8hsuewlah.jpg",
        "partyName": "node package manager",
      },
    ],
    "type": "SET_POLITICAL_PARTIES",
  },
]
`;
